File: battleships/BtlShip.py
# ------------------------------------
# CLASS BtlShip                      |
#                                    |
# Maintain state of an individual    |
# ship                               |
# ------------------------------------
class BtlShip( object ):

  def __init__( self, name, size ):
    self.__name = str( name )
    self.__size = int( size )
    self.__hits = [ 0 ] * self.__size
    self.__x    = None
    self.__y    = None
    self.__dir  = None

  def getName( self ):
    return self.__name

  def getDirection( self ):
    return self.__dir

  def getX( self ):
    return self.__x

  def getY( self ):
    return self.__y

  def getSize( self ):
    return self.__size

  # Set the position of the ship and the direction on the board
  # Is is assumed that this position fits within the current game
  # board, checking should be done in the calling class as BtlShip
  # has no notion of what a board is
  def setPosition( self, x, y, direction ):
    if direction in [ 'v', '>' ]:
      self.__x   = int( x )
      self.__y   = int( y )
      self.__dir = direction

  # Set a section of the ship as hit
  def setHit( self, pos ):
    if pos < self.__size:
      self.__hits[ pos ] = 1

  # Check if part of the ship is at x,y position
  # If shot is True it is record as an actual hit on the ship
  # Does not care or know if position is repeated
  # It is up to class supplying the position to make
  # such checks
  def isHit( self, x, y, shot ):
    if ( self.__x is not None and 
         self.__y is not None and
         self.__dir is not None ):
      # Vertical ship
      if self.__dir == 'v' and self.__x == x:
         if y >= self.__y and y < ( self.__y + self.__size ):
          if shot is True:
            self.__hits[ self.__y - y ] = 1
          return True
      # Horizontal ship
      elif self.__dir == '>' and self.__y == y:
        if x >= self.__x and x < ( self.__x + self.__size ):
          if shot is True:
            self.__hits[ self.__x - x ] = 1
          return True
    return False

  # Check if the ship is sunk
  def isSunk( self ):
    if sum( self.__hits ) == self.__size:
      return True
    else:
      return False
# ------------------------------------

File: javascript/battleships/BtlShip.js
/******************************************************************************
 * Maintains state of a single ship                                          */
function BtlShip( name, size, img ) {
  this.name = name;
  this.size = size;
  this.img = img;
  this.x = null;
  this.y = null;
  this.dir - null;
  this.hits = [];
  for ( var i = 0 ; i < size ; ++i ) {
    this.hits.push( 0 );
  }
}
BtlShip.prototype.getName = function() { return this.name; };
BtlShip.prototype.getDirection = function() { return this.dir; };
BtlShip.prototype.getX = function() { return this.x; };
BtlShip.prototype.getY = function() { return this.y; };
BtlShip.prototype.getSize = function() { return this.size; };
// Record the position of the ship on the board, it is assumed that this
// position fits inside the current board, the checking should be done
// by the callee
BtlShip.prototype.setPosition = function( x, y, direction ) {
  if ( $.inArray( direction, [ 'v', '>' ] ) > -1 ) {
     this.x = x;
     this.y = y;
     this.dir = direction;
  }
};
BtlShip.prototype.setHit = function( pos ) {
  if ( pos < this.size ) {
    this.hits[ pos ] = 1;
  }
};
// Check if part of the ship is at x,y
// If shot is true it will be recorded as a hit
BtlShip.prototype.isHit = function( x, y, shot ) {
  if ( this.x != null && this.y != null && this.dir != null ) {
    // Vertical ship
    if ( this.dir == 'v' && this.x == x ) {
      if ( y >= this.y && y < ( this.y + this.size ) ) {
        if ( shot ) this.hits[ y - this.y ] = 1;
        return true;
      }
    // Horizontal ship
    } else if ( this.dir == '>' && this.y == y ) {
      if ( x >= this.x && x < ( this.x + this.size ) ) {
        if ( shot ) this.hits[ x - this.x ] = 1;
        return true;
      }
    }
  }
  return false;
};
// If part of the ship is at x,y return the correct img
BtlShip.prototype.getImg = function( x, y ) {
  if ( this.x != null && this.y != null && this.dir != null ) {
    // Vertical ship
    if ( this.dir == 'v' && this.x == x ) {
      if ( y >= this.y && y < ( this.y + this.size ) ) {
        return this.img + ( y - this.y ) + '.png';
      }
    // Horizontal ship
    } else if ( this.dir == '>' && this.y == y ) {
      if ( x >= this.x && x < ( this.x + this.size ) ) {
        return this.img + ( x - this.x ) + '.png';
      }
    }
  }
  return false;
};
// Check if the shit has been sunk
BtlShip.prototype.isSunk = function() {
  var sunk = 0;
  for ( var i = 0 ; i < this.hits.length ; ++i ) {
    sunk += this.hits[ i ];
  }
  if ( sunk == this.size ) return true;
  return false;
};
/*****************************************************************************/
